### ISSUE TITLE
Operator concat throws a ConcurrentModificationException

### ISSUE THREAD

**Issue by pditommaso** [nextflow-io/nextflow#675]
The following snippet: 

```
inputs = input1.concat(input2,input3,input4)
```

may throw the following error stack trace: 

```
java.util.ConcurrentModificationException: null
    at java.util.ArrayList$Itr.checkForComodification(ArrayList.java:901)
    at java.util.ArrayList$Itr.next(ArrayList.java:851)
    at nextflow.Session.destroy(Session.groovy:546)
    at nextflow.script.ScriptRunner.terminate(ScriptRunner.groovy:348)
    at nextflow.script.ScriptRunner.execute(ScriptRunner.groovy:167)
    at nextflow.cli.CmdRun.run(CmdRun.groovy:223)
    at nextflow.cli.Launcher.run(Launcher.groovy:428)
    at nextflow.cli.Launcher.main(Launcher.groovy:582)
```

**Comment by pditommaso** [comment]
@fstrozzi This error is very strange, do you still have a complete script to reproduce it ?

**Comment by fstrozzi** [comment]
The error was triggered when using Channel.fromPath in combination with S3. Each of those channels was retrieving different FastQ files with different prefix, e.g. 

```
input1 = Channel.fromFilePairs("s3://bucket/project_1/*/*.fastq.gz",size: -1)
input2 = Channel.fromFilePairs("s3://bucket/project_2/*/*.fastq.gz",size: -1)
input3 = Channel.fromFilePairs("s3://bucket/project_3/*/*.fastq.gz",size: -1)
input4 = Channel.fromFilePairs("s3://bucket/project_4/*/*.fastq.gz",size: -1)

inputs = input1.concat(input2,input3,input4)
inputs.println()

``` 

This piece of code is sufficient to trigger the error, of course you will need some folders and files to test it against. In this case I retrieve data from 150 samples across the 4 projects, so a total number of approximately 350 files. The ```println()``` of the inputs shows a number of results and then returns that error. The error is triggered even if no ```println()``` is present and it is not triggered if the ```concat``` command is used as input for a process, i.e. 

```
process mapping {

  input:
  val data from input1.concat(input2,input3,input4)

  """
  echo $data
  """

}
```

The ```nextflow info``` output is the following:

```
  Version: 0.28.1 build 4781
  Modified: 27-03-2018 16:26 UTC (18:26 CEST)
  System: Mac OS X 10.11.6
  Runtime: Groovy 2.4.13 on OpenJDK 64-Bit Server VM 1.8.0_121-b15
  Encoding: UTF-8 (UTF-8)
```

**Comment by pditommaso** [comment]
Interesting in only happens with S3 paths (!)

**Comment by pditommaso** [comment]
Released in version `0.29.1`.
